rust/src/type_aliases.rs

```diff
@@ -0,0 +1,16 @@
+pub type RequestId = String;
```

Comment:
hi, casual noob here. should these all be [opaque types](https://rustc-dev-guide.rust-lang.org/opaque-types-type-alias-impl-trait.html)? if i were writing this in Scala 3, that's what i'd be doing, but rust != Scala

Reply:
and, to emphasize the context of this question: something like an opaque type lets us leverage the type system for stronger API guarantees on methods like this:

```rust
pub type PassengerId = String;
pub type VehicleId = String;
fn board_vehicle(passenger_id: PassengerId, vehicle_id: VehicleId)
```
where it's easy to mistakenly flip the order of the arguments and the compiler won't notice, versus

```rust
// this may be the correct way to define opaque types but i don't know for sure
pub type PassengerId = impl String; 
pub type VehicleId = impl String;
fn board_vehicle(passenger_id: PassengerId, vehicle_id: VehicleId)
```

where the compiler barks at us if we pass

Reply:
I'm definitely not a Rust expert or a Scala expert (especially Scala 3). But that's a good callout. Based on my reading of [SIP-35](https://docs.scala-lang.org/sips/opaque-types.html), the behavior that Scala opaque types implement:
> In some cases, these authors may believe that using type aliases such as Id and Password means that if they later mix these values up, the compiler will catch their error. However, since type aliases are replaced by their underlying type (e.g. String), these values are considered interchangeable (i.e. type aliases are not appropriate for differentiating various String values).

Is the the Rust [new type idiom](https://doc.rust-lang.org/rust-by-example/generics/new_types.html). They're a bit awkward to use because you either need to dereference or access the tuple value with `.0`. But the benefits are absolutely there. I'd definitely be down to switch everything to this style (and update the existing implementations)

Reply:
We talked about this on site, I'm going to give it a go. It's better to make the switch early

Reply:
rad. thanks for looking into that! the tuple accessor is weird, i'm not sure if that would make HIVE harder to grok for newcomers. but it has it's charm being a zero-cost abstraction (according to [here](https://doc.rust-lang.org/rust-by-example/generics/new_types.html)). so, what's the story with [this](https://rustc-dev-guide.rust-lang.org/opaque-types-type-alias-impl-trait.html) then?